packages/apollo-server-hapi/src/hapiApollo.ts

```diff
@@ -34,56 +39,100 @@ const graphqlHapi: IPlugin = {
     if (!options || !options.graphqlOptions) {
       throw new Error('Apollo Server requires options.');
     }
-    server.route({
-      method: ['GET', 'POST'],
-      path: options.path || '/graphql',
-      vhost: options.vhost || undefined,
-      options: options.route || {},
-      handler: async (request, h) => {
-        try {
-          const { graphqlResponse, responseInit } = await runHttpQuery(
-            [request, h],
-            {
-              method: request.method.toUpperCase(),
-              options: options.graphqlOptions,
-              query:
-                request.method === 'post'
-                  ? // TODO type payload as string or Record
-                    (request.payload as any)
-                  : request.query,
-              request: convertNodeHttpToRequest(request.raw.req),
-            },
-          );
 
-          const response = h.response(graphqlResponse);
-          Object.keys(responseInit.headers).forEach(key =>
-            response.header(key, responseInit.headers[key]),
-          );
-          return response;
-        } catch (error) {
-          if ('HttpQueryError' !== error.name) {
-            throw Boom.boomify(error);
-          }
+    const isUploadEnabled =
+      options.uploadsConfig && typeof processFileUploads === 'function';
+    const path = options.path || '/graphql';
 
-          if (true === error.isGraphQLError) {
-            const response = h.response(error.message);
-            response.code(error.statusCode);
-            response.type('application/json');
-            return response;
-          }
+    const handler: ServerRoute['handler'] = async (request, h) => {
+      try {
+        let query: HttpQueryRequest;
 
-          const err = new Boom(error.message, { statusCode: error.statusCode });
-          if (error.headers) {
-            Object.keys(error.headers).forEach(header => {
-              err.output.headers[header] = error.headers[header];
-            });
-          }
-          // Boom hides the error when status code is 500
-          err.output.payload.message = error.message;
-          throw err;
+        if (request.method === 'get') {
+          query = request.query as any;
+        } else if (isUploadEnabled && request.mime === 'multipart/form-data') {
+          query = (await processFileUploads(
+            request.raw.req,
+            request.raw.res,
+            options.uploadsConfig,
+          )) as any;
+        } else {
+          // The request is unparsed due to the 'POST' route's config. Use
+          // Subtext, hapi's default parser, to parse the request.
+          // https://github.com/hapijs/subtext
+          const { payload } = await Subtext.parse(request.raw.req, null, {
```

Comment:
Subtext does the route payload parsing in hapi: https://github.com/hapijs/hapi/blob/03ca9133d016945ecc8642cc73fa7b05864ec605/lib/route.js#L429

hapi users will already have this library installed.